snap-seccomp, osutil: use osutil.AtomicFile in snap-seccomp

mvo5 opened the pull request:
This PR adds a new osutil.AtomicFile.BackingFile() function that exposes the underlying file. This is then used in snap-seccomp to write the generated bpf in an atomic way.

It also switches the default for tests to snapdUnsafeIO=true and ensures the tests for the AtomicWriteTestSuite are run with both unsafe and safe IO.

I can split this PR up if you think that makes the review easier.

niemeyer commented on cmd/snap-seccomp/main.go:
```diff
@@ -790,27 +789,16 @@ func compile(content []byte, out string) error {
-	}
-	if err := os.Rename(out+".tmp", out); err != nil {
+
+	if err := secFilter.ExportBPF(fout.BackingFile()); err != nil {
```
The new method doesn't seem to be doing much good here. Why not using fout directly?

niemeyer replied:
Not that fout itself (or fout.File) _are_ the backing file.

niemeyer replied:
Note that fout itself (and fout.File) _are_ the backing file.

mvo5 replied:
The way osutil.NewAtomicFile() is written currently does not allow me to use it directly (or fout.File). The reason is that NewAtomicFile returns on interface of AtomicWriter not osutil.atomicFile. However that can be easily changed of course.

Details with the current code:
```
with just fout:
./main.go:798: cannot use fout (type osutil.AtomicWriter) as type *os.File in argument to secFilter.ExportBPF
with fout.File:
./main.go:798: fout.File undefined (type osutil.AtomicWriter has no field or method File)
```

I can of course tweak osutil/io.go and expose the osutil.AtomicFile itself, something like http://paste.ubuntu.com/25525999/ (plus updating all the tests).

mvo5 commented:
@chipaca Thanks a bunch for your feedback. I updated the code to follow my pastebin proposal. I think I added appropriate documentation but I would value feedback on this.

chipaca commented:
Is good. I'd flip-flopped about it being an interface or not originally.

pedronis commented on osutil/export_test.go:
```diff
@@ -93,13 +93,13 @@ func MockChown(f func(*os.File, int, int) error) func() {
 
-func GetAtomicFile(aw AtomicWriter) *os.File {
-	return aw.(*atomicFile).File
+func GetAtomicFile(aw AtomicFile) *os.File {
```
we shouldn't need this one at all now, no?
